File: tests/test_configuration.py
# This file is part of GreenMoon2D, a 2D game engine for Python
# written by Willi Kappler, MIT license.
#
# See: https://github.com/willi-kappler/green_moon_2d

import unittest
import pathlib

from green_moon_2d.gm_configuration import GMConfiguration


class TestConfiguration(unittest.TestCase):
    def create_and_load_config(self, data: str) -> GMConfiguration:
        config_file_name: str = "test_config1.json"

        with open(config_file_name, "w") as f:
            f.write(data)

        cfg: GMConfiguration = GMConfiguration()
        cfg.load_config(config_file_name)
        p = pathlib.Path(config_file_name)
        p.unlink()

        return cfg

    def test_load_configuration1(self):
        """
        Test loading of all configuration values.
        """

        data = """
        {
          "screen_width": 678,
          "screen_height": 456,
          "fullscreen": true,
          "window_title": "Test1",
          "fps": 77,
          "resource_file": "test_resources.json"
        }
        """

        cfg = self.create_and_load_config(data)

        self.assertEqual(cfg.screen_width, 678)
        self.assertEqual(cfg.screen_height, 456)
        self.assertEqual(cfg.fullscreen, True)
        self.assertEqual(cfg.window_title, "Test1")
        self.assertEqual(cfg.fps, 77)
        self.assertEqual(cfg.resource_file, "test_resources.json")

    def test_load_configuration2(self):
        """
        Test loading of no configuration values.
        """

        data = """
        {
          "foo": "bar"
        }
        """

        cfg = self.create_and_load_config(data)

        self.assertEqual(cfg.screen_width, 800)
        self.assertEqual(cfg.screen_height, 600)
        self.assertEqual(cfg.fullscreen, False)
        self.assertEqual(cfg.window_title, "Made with GreenMoon2D")
        self.assertEqual(cfg.fps, 60)
        self.assertEqual(cfg.resource_file, "resources.json")

    def test_load_configuration3(self):
        """
        Test loading of some configuration values.
        """

        data = """
        {
          "screen_width": 1200,
          "resource_file": "foo_bar1.json"
        }
        """

        cfg = self.create_and_load_config(data)

        self.assertEqual(cfg.screen_width, 1200)
        self.assertEqual(cfg.screen_height, 600)
        self.assertEqual(cfg.fullscreen, False)
        self.assertEqual(cfg.window_title, "Made with GreenMoon2D")
        self.assertEqual(cfg.fps, 60)
        self.assertEqual(cfg.resource_file, "foo_bar1.json")


if __name__ == "__main__":
    unittest.main()

File: tests/test_configuration.cpp
/*
    GreenMoon2D
    SPDX-License-Identifier: MIT
    Written by Willi Kappler, MIT License
    https://github.com/willi-kappler/green_moon_2d

    This file contains the tests for the configuration class.

    Run only configuration tests:
    xmake run -w ./ gm2d_test [configuration]
*/

// External includes:
#include <snitch/snitch.hpp>

// Local includes:
#include "gm2d/gm_configuration.hpp"

using namespace gm2d;

TEST_CASE("Create default configuration", "[configuration]" ) {
    GMConfiguration config1;
    REQUIRE(config1.config_file == "config.json");
    REQUIRE(config1.fps == 60);
    REQUIRE(config1.fullscreen == false);
    REQUIRE(config1.resource_file == "resources.json");
    REQUIRE(config1.screen_width == 800);
    REQUIRE(config1.screen_height == 600);
    REQUIRE(config1.window_title == "Made with GreenMoon2D");
}

TEST_CASE("Create configuration from string", "[configuration]" ) {
    std::string config_string = R"json({
        "config_file": "foo_bar_config.json",
        "fps": 70,
        "fullscreen": true,
        "resource_file": "xxx_resource.json",
        "screen_width": 1024,
        "screen_height": 768,
        "window_title": "Test_Foo_Bar"
        })json";

    GMConfiguration config1 = gm_config_from_string(config_string);
    REQUIRE(config1.config_file == "config.json");
    REQUIRE(config1.fps == 70);
    REQUIRE(config1.fullscreen == true);
    REQUIRE(config1.resource_file == "xxx_resource.json");
    REQUIRE(config1.screen_width == 1024);
    REQUIRE(config1.screen_height == 768);
    REQUIRE(config1.window_title == "Test_Foo_Bar");
}
